src/tfp_backend/Code_gen.ml

```diff
@@ -0,0 +1,231 @@
+    (contains_var_expr is_vident)
+    (contains_var_stmt is_vident)
+    accum stmt
+
+let get_param_st p var =
+  let {out_constrained_st= st; _} =
+    List.Assoc.find_exn ~equal:( = ) p.output_vars (get_vident_exn var)
+  in
+  st
+
+let rec get_dims = function
+  | SInt | SReal -> []
+  | SVector d | SRowVector d -> [d]
+  | SMatrix (dim1, dim2) -> [dim1; dim2]
+  | SArray (t, dim) -> dim :: get_dims t
+
+let pp_log_prob ppf p =
+  pf ppf "@ %a@ " pp_log_prob_one_chain p ;
+  let intro =
+    ["return tf__.vectorized_map(self.log_prob_one_chain, params)"]
+  in
+  pp_method ppf "log_prob" ["self"; "params"] intro (fun _ -> ())
+
+let get_params p =
+  List.filter
+    ~f:(function _, {out_block= Parameters; _} -> true | _ -> false)
+    p.output_vars
+
+let pp_shapes ppf p =
+  let pp_shape ppf (_, {out_unconstrained_st; _}) =
+    pf ppf "(nchains__, @[<hov>%a@])" (list ~sep:comma pp_expr)
+      (get_dims out_unconstrained_st)
+  in
+  let ppbody ppf =
+    pf ppf "%a@ " pp_extract_data p ;
+    pf ppf "return [@[<hov>%a@]]" (list ~sep:comma pp_shape) (get_params p)
+  in
+  pp_method ppf "parameter_shapes" ["self"; "nchains__"] [] ppbody
+
+let pp_bijector ppf trans =
+  let pp_call_expr ppf (name, args) = pp_call ppf (name, pp_expr, args) in
+  let components =
+    match trans with
+    | Identity -> []
+    | Lower lb -> [("Exp", []); ("AffineScalar", [lb])]
+    | _ ->
+        raise_s
+          [%message "Unsupported " (trans : expr_typed_located transformation)]
+  in
+  match components with
+  | [] -> pf ppf "tfb__.Identity()"
+  | ls ->
+      pf ppf "tfb__.Chain([@[<hov>%a@]])"
+        (list ~sep:comma pp_call_expr)
+        List.(rev (map ls ~f:(fun (s, args) -> ("tfb__." ^ s, args))))
+
+let pp_bijectors ppf p =
+  let ppbody ppf =
+    pf ppf "%a@ " pp_extract_data p ;
+    pf ppf "return [@[<hov>%a@]]"
+      (list ~sep:comma pp_bijector)
+      (List.map ~f:(fun (_, {out_trans; _}) -> out_trans) (get_params p))
+  in
+  pp_method ppf "parameter_bijectors" ["self"] [] ppbody
+
+let pp_methods ppf p =
+  pf ppf "@ %a" pp_init p ;
+  pf ppf "@ %a" pp_log_prob p ;
+  pf ppf "@ %a" pp_shapes p ;
+  pf ppf "@ %a" pp_bijectors p
+
+let pp_fundef ppf {fdname; fdargs; fdbody; _} =
+  pp_method ppf fdname
+    (List.map ~f:(fun (_, name, _) -> name) fdargs)
+    []
+    (fun ppf -> pp_stmt ppf fdbody)
+
+let imports =
+  {|
+import numpy as np__
+import tensorflow as tf__
+import tensorflow_probability as tfp__
+tfd__ = tfp.distributions
+tfb__ = tfp.bijectors
+from tensorflow.python.ops.parallel_for import pfor as pfor__
+|}
+
+let pp_prog ppf (p : typed_prog) =
+  pf ppf "%s@,@,%a@,class %s(tfd__.Distribution):@,@[<v 2>%a@]" imports
+    (list ~sep:cut pp_fundef) p.functions_block p.prog_name pp_methods p
+
+(* Major work to do:
+1. Work awareness of distributions and bijectors into the type system
+2. Have backends present an environment that the frontend and middle can use for type checking and optimization.
+*)
```

Comment:
This is sort of a meta-comment but I think there may be some value in looking at translating to a TFP DSL before generating raw python. It's pretty tricky to reason about generating strings vs. a principled translation to another data structure.

Reply:
We talked about this in a meeting - at some point you end up with strings and in many ways the MIR is that low level DSL (with a TFP specific Transform_mir step happening first). Leaving this for future innovation once we figure out what we want the MIR to solve.